geoutils/raster/delayed.py

```diff
@@ -0,0 +1,777 @@
+        resampling=kwargs["resampling"],
+        src_nodata=kwargs["src_nodata"],
+        dst_nodata=kwargs["dst_nodata"],
+    )
+
+    return dst_arr
+
+
+def delayed_reproject(
+    darr: da.Array,
+    src_transform: rio.transform.Affine,
+    src_crs: rio.crs.CRS,
+    dst_transform: rio.transform.Affine,
+    dst_shape: tuple[int, int],
+    dst_crs: rio.crs.CRS,
+    resampling: rio.enums.Resampling,
+    src_nodata: int | float | None = None,
+    dst_nodata: int | float | None = None,
+    dst_chunksizes: tuple[int, int] | None = None,
+    **kwargs: Any,
+) -> da.Array:
+    """
+    Reproject georeferenced raster on out-of-memory chunks.
+
+    Each chunk of the destination array is mapped to one or several intersecting chunks of the source array, and
+    reprojection is performed using rio.warp.reproject for each mapping.
+
+    Part of the code is inspired by https://github.com/opendatacube/odc-geo/pull/88.
+
+    :param darr: Input dask array for source raster.
+    :param src_transform: Geotransform of source raster.
+    :param src_crs: Coordinate reference system of source raster.
+    :param dst_transform: Geotransform of destination raster.
+    :param dst_shape: Shape of destination raster.
+    :param dst_crs: Coordinate reference system of destination raster.
+    :param resampling: Resampling method.
+    :param src_nodata: Nodata value of source raster.
+    :param dst_nodata: Nodata value of destination raster.
+    :param dst_chunksizes: Chunksizes for destination raster.
+    :param kwargs: Other arguments to pass to rio.warp.reproject().
+
+    :return: Dask array of reprojected raster.
+    """
+
+    # 1/ Define source and destination chunked georeferenced grid through simple classes storing CRS/transform/shape,
+    # which allow to consistently derive shape/transform for each block and their CRS-projected footprints
+
+    # Define georeferenced grids for source/destination array
+    src_geogrid = GeoGrid(transform=src_transform, shape=darr.shape, crs=src_crs)
+    dst_geogrid = GeoGrid(transform=dst_transform, shape=dst_shape, crs=dst_crs)
+
+    # Add the chunking
+    # For source, we can use the .chunks attribute
+    src_chunks = darr.chunks
+    src_geotiling = ChunkedGeoGrid(grid=src_geogrid, chunks=src_chunks)
+
+    # For destination, we need to create the chunks based on destination chunksizes
+    if dst_chunksizes is None:
+        dst_chunksizes = darr.chunksize
+    dst_chunks = _chunks2d_from_chunksizes_shape(chunksizes=dst_chunksizes, shape=dst_shape)
+    dst_geotiling = ChunkedGeoGrid(grid=dst_geogrid, chunks=dst_chunks)
+
+    # 2/ Get footprints of tiles in CRS of destination array, with a buffer of 2 pixels for destination ones to ensure
+    # overlap, then map indexes of source blocks that intersect a given destination block
+    src_footprints = src_geotiling.get_block_footprints(crs=dst_crs)
+    dst_footprints = dst_geotiling.get_block_footprints().buffer(2 * max(dst_geogrid.res))
```

Comment:
How is the 2 pixels buffer decided ? Is this enough for reprojections?

Reply:
Yes normally no resampling algorithm uses more than +2/-2, this is the same in ODC. We could do more checking on this if the Rasterio reprojection issues were solved... It'll have to wait.